experimental/dds/stack/src/sharedStack.ts

```diff
@@ -0,0 +1,162 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { IEvent } from "@fluidframework/common-definitions";
+import { IsoBuffer } from "@fluidframework/common-utils";
+import { IFluidHandle } from "@fluidframework/core-interfaces";
+import { IChannel, IChannelAttributes, IChannelFactory, IChannelServices, IChannelStorageService, IFluidDataStoreRuntime } from "@fluidframework/datastore-definitions";
+import { ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+import { ITelemetryContext, ISummaryTreeWithStats, IGarbageCollectionData } from "@fluidframework/runtime-definitions";
+import { SummaryTreeBuilder } from "@fluidframework/runtime-utils";
+import { IFluidSerializer, SharedObject } from "@fluidframework/shared-object-base";
+
+/**
+ * Events useful for testing and debugging a {@link SharedStack}
+ */
+export interface DiagnosticEvents extends IEvent {
+    (event: "downloadedBlob", handler: () => void);
+    (event: "uploadedBlob", handler: () => void);
+    (event: "processedOp", handler: () => void);
+}
+
+/**
+ * A stack DDS which stores every element in its own attachment blob.
+ */
+export class SharedStack<T = unknown> extends SharedObject<DiagnosticEvents> {
+    private head?: IFluidHandle<ArrayBufferLike>;
+
+    /** Construct a new SharedStack */
+    public constructor(
+        id: string,
+        runtime: IFluidDataStoreRuntime,
+        attributes: IChannelAttributes,
+        telemetryContextPrefix: string
+    ) {
+        super(id, runtime, attributes, telemetryContextPrefix);
+    }
+
+    /** True iff there are no elements pushed to this stack */
+    public isEmpty(): boolean {
+        return this.head === undefined;
+    }
+
+    /** Download, remove and return the top element on this stack */
+    public async pop(): Promise<T> {
+        if (this.head === undefined) {
+            throw new Error("SharedStack is empty");
+        }
+
+        const head = await get<Node<T>>(this.head, this.serializer);
+        this.emit("downloadedBlob");
+        this.setHead(head.next);
```

Comment:
Since the head is only updated after an await, it looks like a client could call pop() twice without awaiting it and see the same element both times.  Is it safe to update the head prior to running the async portion?  Is it a requirement for the client to `await` the pop() before calling pop() again?

Reply:
These comments are great feedback, thank you. I'll be honest that I didn't think deeply through any of the concurrency scenarios for this DDS, since right now I'm just trying to get a simple testbed for attachment blobs and GC up and running. But let's see here... my first instinct is that if the caller is popping twice without awaiting, then it's fair game to return the same element. If I were eating an ice cream cone with three scoops on it, and someone ordered me to "eat the top scoop!" and then I was only halfway done licking it, and they again said "eat the top scoop!", I would only feel obligated to finish that same scoop. But maybe that's a bad analogy because I'm gonna eat all the scoops no matter what they say anyway :P

Reply:
If this is a weird scenario and the head update has to come after the await, perhaps just blow up ("can't pop while another pop is pending" plus discoverable canEdit() state as mentioned above).  I think it's easier to explain that to the customer than "When you call pop(), you might get the same element that you (or another async task) has received in a previous pop() if that previous pop() has not concluded"

I was debating whether you could queue the push/pop requests internally and only resolve them when safe, but I think it probably gets funky if it starts lagging behind too much.  Maybe worth thinking through a little bit though.